### PR TITLE
[X86][AMX] Support AMX-AVX512

### PR DESCRIPTION
**Pull request by phoebewang**

### CODE REVIEW

File: clang/lib/Headers/amxavx512intrin.h [hunk 1]
```diff
@@ -0,0 +1,381 @@
...
+
+#ifndef __AMX_AVX512INTRIN_H
+#define __AMX_AVX512INTRIN_H
+#ifdef __x86_64__
+
+#define __DEFAULT_FN_ATTRS_AVX512                                              \
+  __attribute__((__always_inline__, __nodebug__, __target__("amx-avx512")))
+
+/// Moves a row from a tile register to a zmm destination register, converting
+///    the int32 source elements to fp32. The row of the tile is selected by an
```
[the remark this answers is not in the document]
Comment (phoebewang) [reply clang/lib/Headers/amxavx512intrin.h]: Done.

File: clang/lib/Headers/amxavx512intrin.h [hunk 2]
```diff
@@ -0,0 +1,381 @@
...
+///     FI
+/// ENDFOR
+/// dst[MAX_VL-1:VL] := 0
+/// zero_tileconfig_start()
+/// \endcode
+///
+/// This intrinsic corresponds to the \c TCVTROWD2PS instruction.
+///
+/// \param tsrc
+///    The 1st source tile. Max size is 1024 Bytes.
```
[the remark this answers is not in the document]
Comment (phoebewang) [reply clang/lib/Headers/amxavx512intrin.h]: Done.

File: clang/lib/Headers/amxavx512intrin.h [hunk 3]
```diff
@@ -0,0 +1,381 @@
...
+/// \param row
+///    The row of the source tile
+#define _tile_cvtrowd2ps(tsrc, row) __builtin_ia32_tcvtrowd2ps(tsrc, row)
+
+/// Moves a row from a tile register to a zmm destination register, converting
+///    the fp32 source elements to bf16. It places the resulting bf16 elements
+///    in the high 16 bits within each dword. The row of the tile is selected
+///    by an 32b GPR.
```
[the remark this answers is not in the document]
Comment (phoebewang) [reply clang/lib/Headers/amxavx512intrin.h]: Done.

File: llvm/lib/Target/X86/X86ExpandPseudo.cpp [hunk 4]
```diff
@@ -559,12 +559,68 @@ bool X86ExpandPseudo::expandMI(MachineBasicBlock &MBB,
...
+      Opc = X86::TCVTROWPS2PHLrri;
+      break;
+    case X86::PTILEMOVROWrreV:
+      Opc = X86::TILEMOVROWrre;
+      break;
+    case X86::PTILEMOVROWrriV:
+      Opc = X86::TILEMOVROWrri;
+      break;
+    default:
+      llvm_unreachable("Impossible Opcode!");
```
[the remark this answers is not in the document]
Comment (phoebewang) [reply llvm/lib/Target/X86/X86ExpandPseudo.cpp]: Done.

File: clang/include/clang/Basic/BuiltinsX86_64.def [hunk 5]
```diff
@@ -133,6 +133,12 @@ TARGET_BUILTIN(__builtin_ia32_t2rpntlvwz0t1_internal, "vUsUsUsV256i*V256i*vC*z",
 TARGET_BUILTIN(__builtin_ia32_t2rpntlvwz1_internal, "vUsUsUsV256i*V256i*vC*z", "n", "amx-transpose")
 TARGET_BUILTIN(__builtin_ia32_t2rpntlvwz1t1_internal, "vUsUsUsV256i*V256i*vC*z", "n", "amx-transpose")
 TARGET_BUILTIN(__builtin_ia32_ttransposed_internal, "V256iUsUsV256i", "n", "amx-transpose")
+TARGET_BUILTIN(__builtin_ia32_tcvtrowd2ps_internal, "V16fUsUsV256iUi", "n", "amx-avx512")
```
[the remark this answers is not in the document]
Comment (phoebewang) [reply clang/include/clang/Basic/BuiltinsX86_64.def:136]: Good catch! Done.

[the remark this answers is not in the document]
Comment (fzou1) [reply clang/include/clang/Basic/BuiltinsX86_64.def:136]: Is it necessary to add avx10.2-512 feature for these internal APIs? With that, we may detect errors for new APIs if there is no AVX10.2-512 target feature.

File: clang/lib/Headers/amxavx512intrin.h [hunk 6]
```diff
@@ -0,0 +1,381 @@
...
+#ifndef __IMMINTRIN_H
+#error "Never use <amxavx512intrin.h> directly; include <immintrin.h> instead."
+#endif // __IMMINTRIN_H
+
+#ifndef __AMX_AVX512INTRIN_H
+#define __AMX_AVX512INTRIN_H
+#ifdef __x86_64__
+
+#define __DEFAULT_FN_ATTRS_AVX512                                              \
+  __attribute__((__always_inline__, __nodebug__, __target__("amx-avx512")))
```
Comment (fzou1) [inline clang/lib/Headers/amxavx512intrin.h:18]: If AVX10.2-512 feature dependency is needed for internal APIs, we may create another attribute with AVX10.2-512 and add it to internal APIs.

File: clang/include/clang/Basic/BuiltinsX86_64.def (hunk 5 above)
[the remark this answers is not in the document]
Comment (phoebewang) [reply clang/include/clang/Basic/BuiltinsX86_64.def]: Yes, sorry I missed that parts.

File: clang/lib/Headers/amxavx512intrin.h (hunk 6 above)
Comment (phoebewang) [reply clang/lib/Headers/amxavx512intrin.h]: Yes, we need them all. Good catch!

File: llvm/lib/Target/X86/X86InstrAMX.td [hunk 7]
```diff
@@ -369,3 +369,150 @@ let Predicates = [HasAMXTRANSPOSE, In64BitMode] in {
     }
   }
 } // HasAMXTILE, HasAMXTRANSPOSE
+
+multiclass m_tcvtrowd2ps {
+  let Predicates = [HasAMXAVX512, In64BitMode] in {
```
Comment (fzou1) [inline llvm/lib/Target/X86/X86InstrAMX.td:374]: Should add HasAVX10_2_512 in line 374, 390 and 454?

**Review by fzou1 — COMMENTED** [review]
LGTM except the last place probably missing avx10.2-512 dependency.

File: llvm/lib/Target/X86/X86InstrAMX.td (hunk 7 above)
Comment (phoebewang) [reply llvm/lib/Target/X86/X86InstrAMX.td]: Done, thanks!

**Review by fzou1 — APPROVED** [review]
LGTM

**Comment by alanzhao1** [comment]
FYI this is causing Chrome X86 MacOS builds to fail due to `error: unknown type name '__m512bh'`: https://crbug.com/378111077

**Comment by alanzhao1** [comment]
> FYI this is causing Chrome X86 MacOS builds to fail due to `error: unknown type name '__m512bh'`: https://crbug.com/378111077

As I mentioned in https://crbug.com/378111077#comment3, the issue is that we pull in avx512bf16intrin.h because `__SCE__` is not defined, but `__SSE2__` is also not defined, so we don't pull in the `typedef` for `__m512bh` (since they're guarded by macros that look for `__SSE2__`) and so on. We then pull in amxavx512intrin.h which was introduced in this PR which then fails to compile because we don't have the `typedef`s.

**Comment by pranavk** [comment]
This is causing similar errors for us as well.

**Comment by alanzhao1** [comment]
Figured out a repro - `immintrin.h` doesn't compile if both `-msse` and `-mno-sse2` are passed to clang:

```
$ cat ~/src/test-mac.c
#include<immintrin.h>

$ bin/clang -msse -mno-sse2 -o /dev/null -c ~/src/test-mac.c
In file included from /usr/local/google/home/ayzhao/src/test-mac.c:1:
In file included from /usr/local/google/home/ayzhao/src/llvm-project/build/lib/clang/20/include/immintrin.h:660:
/usr/local/google/home/ayzhao/src/llvm-project/build/lib/clang/20/include/amxavx512intrin.h:240:19: error: unknown type name '__m512bh'
  240 | static __inline__ __m512bh __DEFAULT_FN_ATTRS_AVX512
      |                   ^
/usr/local/google/home/ayzhao/src/llvm-project/build/lib/clang/20/include/amxavx512intrin.h:243:10: error: returning '__attribute__((__vector_size__(32 * sizeof(__bf16)))) __bf16' (vector of 32 '__bf16' values) from a function with incompatible result type 'int'
  243 |   return __builtin_ia32_tcvtrowps2pbf16h_internal(m, n, src, u);
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/local/google/home/ayzhao/src/llvm-project/build/lib/clang/20/include/amxavx512intrin.h:246:19: error: unknown type name '__m512bh'
  246 | static __inline__ __m512bh __DEFAULT_FN_ATTRS_AVX512
      |                   ^
/usr/local/google/home/ayzhao/src/llvm-project/build/lib/clang/20/include/amxavx512intrin.h:249:10: error: returning '__attribute__((__vector_size__(32 * sizeof(__bf16)))) __bf16' (vector of 32 '__bf16' values) from a function with incompatible result type 'int'
  249 |   return __builtin_ia32_tcvtrowps2pbf16l_internal(m, n, src, u);
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/local/google/home/ayzhao/src/llvm-project/build/lib/clang/20/include/amxavx512intrin.h:252:19: error: unknown type name '__m512h'
  252 | static __inline__ __m512h __DEFAULT_FN_ATTRS_AVX512 _tile_cvtrowps2phh_internal(
      |                   ^
/usr/local/google/home/ayzhao/src/llvm-project/build/lib/clang/20/include/amxavx512intrin.h:254:10: error: returning '__attribute__((__vector_size__(32 * sizeof(_Float16)))) _Float16' (vector of 32 '_Float16' values) from a function with incompatible result type 'int'
  254 |   return __builtin_ia32_tcvtrowps2phh_internal(m, n, src, u);
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/local/google/home/ayzhao/src/llvm-project/build/lib/clang/20/include/amxavx512intrin.h:257:19: error: unknown type name '__m512h'
  257 | static __inline__ __m512h __DEFAULT_FN_ATTRS_AVX512 _tile_cvtrowps2phl_internal(
      |                   ^
/usr/local/google/home/ayzhao/src/llvm-project/build/lib/clang/20/include/amxavx512intrin.h:259:10: error: returning '__attribute__((__vector_size__(32 * sizeof(_Float16)))) _Float16' (vector of 32 '_Float16' values) from a function with incompatible result type 'int'
  259 |   return __builtin_ia32_tcvtrowps2phl_internal(m, n, src, u);
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/local/google/home/ayzhao/src/llvm-project/build/lib/clang/20/include/amxavx512intrin.h:302:8: error: unknown type name '__m512bh'
  302 | static __m512bh __tile_cvtrowps2pbf16h(__tile1024i src0, unsigned src1) {
      |        ^
/usr/local/google/home/ayzhao/src/llvm-project/build/lib/clang/20/include/amxavx512intrin.h:321:8: error: unknown type name '__m512bh'
  321 | static __m512bh __tile_cvtrowps2pbf16l(__tile1024i src0, unsigned src1) {
      |        ^
/usr/local/google/home/ayzhao/src/llvm-project/build/lib/clang/20/include/amxavx512intrin.h:340:8: error: unknown type name '__m512h'
  340 | static __m512h __tile_cvtrowps2phh(__tile1024i src0, unsigned src1) {
      |        ^
/usr/local/google/home/ayzhao/src/llvm-project/build/lib/clang/20/include/amxavx512intrin.h:359:8: error: unknown type name '__m512h'
  359 | static __m512h __tile_cvtrowps2phl(__tile1024i src0, unsigned src1) {
      |        ^
12 errors generated.
```

Test was on an x64 linux system.

**Comment by phoebewang** [comment]
Thanks all! Fixed in https://github.com/llvm/llvm-project/pull/115581

**Pull request merged**